### PR TITLE
Feature/csm 169 extend search epoch

### PR DESCRIPTION
**Pull request by ksaric**
This should be both:
- https://issues.serokell.io/issue/CSM-169
- https://issues.serokell.io/issue/CSM-170

### CODE REVIEW

**Comment by ksaric** [comment]
Tested with:

`GET http://localhost:8100/api/search/epoch/4`

```json
{"Right":[{"cbeEpoch":4,"cbeSlot":0,"cbeBlkHash":"5ae7775556c2f46d06cf157cbb0ca2c9e37805d0218b6e764fdaa9ebbf3da097","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":1,"cbeBlkHash":"d277135806d4f9650f31a75f9be1394e170fa9080833048d211e19b068711ace","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":2538,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":2,"cbeBlkHash":"944e2da2c66b2f4d63cf13e33e3e5def6243b70ea15c919ba51898ab264248a7","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":3,"cbeBlkHash":"a022878acef4b39237ddc989889a870e5d216d3013e74ffd1b89e82e2747ee13","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":4,"cbeBlkHash":"31d2bedefe8a1295f1c85ee459b951f8567298c4e1e6b20b3ec034c367851964","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":387,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":5,"cbeBlkHash":"98a3279ea12c72c5a5204d64b3c2bc42f9377f69d463f20e5d716e44c303beb2","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":387,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":6,"cbeBlkHash":"cc28b5e1a7b2afbce8f4b00c05aa15055f2894790ee3fcfda22e320747af22a0","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":387,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":7,"cbeBlkHash":"47811ba4d95e3222947afe9982425486b897176835f84fa3dc23d2945c3311fe","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":387,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":9,"cbeBlkHash":"51b62680dad4a8a585b724666c3e7bb11f2bb32ed6311a35b7f743412719d66a","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":542,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":10,"cbeBlkHash":"d4b32a5e88b0ce76e1e2316df3ed70b39a526c42a9cd0812bca9211acab35622","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":12,"cbeBlkHash":"20c00a20e589ff5420a4d1d25e20b71030fdfaff2299005f90ebc334444f620c","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":387,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":13,"cbeBlkHash":"5367fc1b38bd1463ed11c2f0af21c5c6a76da27f265ff5c291b9f5aa35e72d0c","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":387,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":14,"cbeBlkHash":"9d36fbc06df2aaa9c7bf80164b9c340c99997d7c8380794bdfa21c7aa8f02862","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":387,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":16,"cbeBlkHash":"7ff9a8256b7b47c5d9e29eb29fe43310f3518bafcfa4a30a6c621f972a361e8d","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":738,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":17,"cbeBlkHash":"079f61558a0f0c77fecb3d1eec7183c2a55f8dc8e3dab34233d862b5c3a4fab8","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":18,"cbeBlkHash":"8ade7bbe727f5b6dc2ce44b118184e78a38e6abc23c05fe987447e0cbeebfaa7","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null},{"cbeEpoch":4,"cbeSlot":19,"cbeBlkHash":"bfcc2b1270382de44051fdcf7148b7104e535b7cfc29ffb39fc0a9bbe5b05a75","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null}]}
```

`GET http://localhost:8100/api/search/epoch/4?slot=2`

```json
{"Right":[{"cbeEpoch":4,"cbeSlot":2,"cbeBlkHash":"944e2da2c66b2f4d63cf13e33e3e5def6243b70ea15c919ba51898ab264248a7","cbeTimeIssued":null,"cbeTxNum":0,"cbeTotalSent":{"getCoin":0},"cbeSize":420,"cbeRelayedBy":null}]}
```

**Comment by ksaric** [comment]
Also, @akegalj, this needs to be pushed to `master` also, right?

**Comment by akegalj** [comment]
@ksaric 
> Also, @akegalj, this needs to be pushed to master also, right?
In general, yes. But before new release `0.4` we will probably keep `master` and `0.3` in sync so its ok.

File: src/Pos/Explorer/Web/Server.hs [hunk 1]
```diff
@@ -250,6 +304,45 @@ getTxSummary cTxId = do
...
 
+-- | Find all `MainBlock` by applying the *predicate*, starting from *headerHash*
+filterMainBlocks :: (SscHelpersClass ssc, ExplorerMode m) =>
+    HeaderHash ->
+    (MainBlock ssc -> Bool) ->
+    m [MainBlock ssc]
+filterMainBlocks headerHash predicate = rights <$> generalBlockSearch
+  where
+    generalBlockSearch = filterAllBlocks headerHash specializedPred (pure [])
+    specializedPred block = case block of
```
Comment (akegalj) [inline src/Pos/Explorer/Web/Server.hs]: I would implement it as:
```
either (const False) predicate
```

File: src/Pos/Explorer/Web/Server.hs [hunk 2]
```diff
@@ -250,6 +304,45 @@ getTxSummary cTxId = do
...
+    -- blocks (hash) to accumulator satisfying the predicate.
+    | otherwise = do
+        -- Get the block with the sent hash, throw exception if/when the block
+        -- search fails.
+        block <- DB.getBlock headerHash >>=
+            maybeThrow (Internal "Block with hash cannot be found!")
+        -- If there is a block then iterate backwards with the predicate
+        let prevBlock = block ^. prevBlockL
+
+        case predicate block of
```
Comment (akegalj) [inline src/Pos/Explorer/Web/Server.hs]: Use `if` here:
```
if predicate block
   then a
   else b
```

File: src/Pos/Explorer/Web/Server.hs [hunk 3]
```diff
@@ -250,6 +304,45 @@ getTxSummary cTxId = do
...
+-- | Find all blocks matching the sent predicate. This is a generic function
+-- that can be called with either `MainBlock` or `GenesisBlock` in mind.
+filterAllBlocks :: (SscHelpersClass ssc, ExplorerMode m) =>
+    HeaderHash ->
+    (Block ssc -> Bool) ->
+    m [Block ssc] ->
+    m [Block ssc]
+filterAllBlocks headerHash predicate acc
+    -- When we reach the genesis block, return the accumulator.
+    | headerHash == genesisHash = acc
```
Comment (akegalj) [inline src/Pos/Explorer/Web/Server.hs]: Would also prefer `if` here as it has only two branches:
```
if headerHash == genesis
  then acc
  else rest
```

File: src/Pos/Explorer/Web/Server.hs [hunk 4]
```diff
@@ -135,9 +157,41 @@ searchHash shash = getResult $ do
...
+    localSlotIndex = fmap LocalSlotIndex slotIndex
+    transformToCBlockEntry blocks = sequence $ toBlockEntry <$> blocks
+
+-- | Get all blocks by epoch and slot. The slot is optional, if it exists,
+-- it just adds another predicate to match it.
+getBlocksByEpoch :: (SscHelpersClass ssc, ExplorerMode m) =>
+    EpochIndex ->
+    Maybe LocalSlotIndex ->
+    m [MainBlock ssc]
+getBlocksByEpoch epochIndex Nothing = do
```
Comment (akegalj) [inline src/Pos/Explorer/Web/Server.hs]: Both of these branches are pretty similar. Try something like:
```
getBlocksByEpoch epochIndex mSlotIndex = do
      tipHash <- GS.getTip
     filterMainBlocks tipHash findBlocksByEpochPred
        findBlocksByEpochPred mb =
            (siEpoch $ mb ^. blockSlot) == epochIndex &&
            fromMaybe True ((== siSlot  $ mb ^. blockSlot) <$> mSlotIndex)
```

File: src/Pos/Explorer/Web/Server.hs (hunk 3 above)
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: Just two branches? :)

Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: btw, I like current way more than `if` one

File: src/Pos/Explorer/Web/Server.hs [hunk 5]
```diff
@@ -135,9 +157,41 @@ searchHash shash = getResult $ do
...
+-- | Search the blocks by epoch and slot. Slot is optional.
+epochSlotSearch :: (ExplorerMode m) =>
+    EpochIndex ->
+    Maybe Word16 ->
+    m [CBlockEntry]
+epochSlotSearch epochIndex slotIndex = findBlocksByEpoch >>= transformToCBlockEntry
+  where
+    findBlocksByEpoch = getBlocksByEpoch @SscGodTossing epochIndex localSlotIndex
+    localSlotIndex = fmap LocalSlotIndex slotIndex
+    transformToCBlockEntry blocks = sequence $ toBlockEntry <$> blocks
```
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: I think `transformToCBlockEntry` = `traverse`

File: src/Pos/Explorer/Web/Server.hs [hunk 6]
```diff
@@ -135,9 +157,41 @@ searchHash shash = getResult $ do
...
+
+-- | Get all blocks by epoch and slot. The slot is optional, if it exists,
+-- it just adds another predicate to match it.
+getBlocksByEpoch :: (SscHelpersClass ssc, ExplorerMode m) =>
```
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: Please, wrap function signatures like here:
https://github.com/input-output-hk/cardano-sl/blob/master/src/Pos/Wallet/Update.hs

File: src/Pos/Explorer/Web/Server.hs [hunk 7]
```diff
@@ -250,6 +304,45 @@ getTxSummary cTxId = do
...
+
+-- | Find all blocks matching the sent predicate. This is a generic function
+-- that can be called with either `MainBlock` or `GenesisBlock` in mind.
+filterAllBlocks :: (SscHelpersClass ssc, ExplorerMode m) =>
+    HeaderHash ->
+    (Block ssc -> Bool) ->
+    m [Block ssc] ->
+    m [Block ssc]
+filterAllBlocks headerHash predicate acc
+    -- When we reach the genesis block, return the accumulator.
```
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: Could you explain why do you stop at genesis block?

File: src/Pos/Explorer/Web/Server.hs (hunk 3 above)
Comment (akegalj) [inline src/Pos/Explorer/Web/Server.hs]: Then we can leave it. Its just my opinion.

File: src/Pos/Explorer/Web/Server.hs (hunk 4 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: Fixed, I find this harder to read.

File: src/Pos/Explorer/Web/Server.hs (hunk 1 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: Fixed.

File: src/Pos/Explorer/Web/Server.hs (hunk 3 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: Leaving guards.

File: src/Pos/Explorer/Web/Server.hs (hunk 2 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: Fixed. I think if should be avoided in favor of case matching which should be idiomatic Haskell, right?

File: src/Pos/Explorer/Web/Server.hs (hunk 5 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: Fixed.

File: src/Pos/Explorer/Web/Server.hs (hunk 6 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: No problem, but the rest of the code should follow the same convention? It's easier to read if all types are aligned the same.

File: src/Pos/Explorer/Web/Server.hs (hunk 7 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: Updated the comment.

File: src/Pos/Explorer/Web/Server.hs (hunk 2 above)
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: IMO If look at `if` as `case` shortcut for `Bool`s, then why not to use it

File: src/Pos/Explorer/Web/Server.hs (hunk 6 above)
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: Which rest of the code do you mean?

File: src/Pos/Explorer/Web/Server.hs (hunk 7 above)
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: Actually I meant, just write it here, but code comment is even better.
Well, I thought you discovered in #explorer that genesis block is present in each *epoch* once, not in blockchain once, or I'm wrong?
https://input-output-rnd.slack.com/archives/C3NJV7ADQ/p1490368930079121

File: src/Pos/Explorer/Web/Server.hs [hunk 8]
```diff
@@ -135,10 +157,36 @@ searchHash shash = getResult $ do
...
+epochSlotSearch
+    :: (ExplorerMode m)
+    => EpochIndex
+    -> Maybe Word16
+    -> m [CBlockEntry]
+epochSlotSearch epochIndex slotIndex = findBlocksByEpoch >>= transformToCBlockEntry
+  where
+    findBlocksByEpoch = getBlocksByEpoch @SscGodTossing epochIndex localSlotIndex
+    localSlotIndex = fmap LocalSlotIndex slotIndex
+    transformToCBlockEntry blocks = traverse toBlockEntry blocks
```
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: If you want to, eta-reducing it would be also nice

File: src/Pos/Explorer/Web/Server.hs [hunk 9]
```diff
@@ -250,6 +298,47 @@ getTxSummary cTxId = do
...
+
+-- | Find all blocks matching the sent predicate. This is a generic function
+-- that can be called with either `MainBlock` or `GenesisBlock` in mind.
+filterAllBlocks
+    :: (SscHelpersClass ssc, ExplorerMode m)
+    => HeaderHash
+    -> (Block ssc -> Bool)
+    -> m [Block ssc]
+    -> m [Block ssc]
+filterAllBlocks headerHash predicate acc
```
Comment (akegalj) [inline src/Pos/Explorer/Web/Server.hs]: Note that manuall recursions tend to cause memory leaks if they are not strict in the accumulator. I will review it in more details tomorrow but usually similar loops are implemented with fold (which is much safer to reason about memory wise then custom recursion)

Will take a closer look tomorrow

File: src/Pos/Explorer/Web/Server.hs [hunk 10]
```diff
@@ -250,6 +298,47 @@ getTxSummary cTxId = do
...
+-- | Find all `MainBlock` by applying the *predicate*, starting from *headerHash*
+filterMainBlocks
+    :: (SscHelpersClass ssc, ExplorerMode m)
+    => HeaderHash
+    -> (MainBlock ssc -> Bool)
+    -> m [MainBlock ssc]
+filterMainBlocks headerHash predicate = rights <$> generalBlockSearch
+  where
+    generalBlockSearch    = filterAllBlocks headerHash specializedPred (pure [])
+    specializedPred block = either (const False) predicate block
```
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: And, perhaps you meant `specialized` -> `specified`?

File: src/Pos/Explorer/Web/Server.hs (hunk 6 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: I mean the rest of the codebase of `cardano-sl-explorer`, it seems `cardano-sl` is (mostly) written in that style.

Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: Could you point to concrete file? Till this moment I was sure in the opposite )

File: src/Pos/Explorer/Web/Server.hs (hunk 10 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: I can change that, but I did mean `specialized`, as in "this predicate is **specialized** for the `MainBlock`".

File: src/Pos/Explorer/Web/Server.hs (hunk 8 above)
Comment (ksaric) [inline src/Pos/Explorer/Web/Server.hs]: Fixed.

File: src/Pos/Explorer/Web/Server.hs (hunk 10 above)
Comment (Martoon-00) [inline src/Pos/Explorer/Web/Server.hs]: Aha, got it, sorry for doubt

**Pull request merged**